lib/rspec/mocks/method_reference.rb

```diff
@@ -94,7 +94,16 @@ def self.method_visibility_for(object, method_name)
           # write `method_missing` in such a way that it handles a dynamic message
           # with private or protected visibility. Ruby doesn't provide you with
           # the caller info.
-          return :public if vis.nil? && object.respond_to?(method_name, false)
+          proxy = RSpec::Mocks.space.proxy_for(object)
+          respond_to = proxy.method_double_if_exists_for_message(:respond_to?)
+
+          if proxy && respond_to
+            object_responds_to_method_name = respond_to.original_method.call(method_name)
+          else
+            object_responds_to_method_name = object.respond_to?(method_name)
+          end
+
+          return :public if vis.nil? && object_responds_to_method_name
         end
```

Comment:
I think the entire method should be rewritten to not use `tap`:

``` ruby
vis = instance_method_visibility_for(class << object; self; end, method_name)
return vis if vis

# the reset of your logic
```

As you've written it, it wastes time trying to get the proxy, method double and original method but we only need to bother if the return value of `instance_method_visibility_for` is `nil`.  The use of `tap` here makes it kinda confusing, too.